### PR TITLE
Add Bert mesh experiments 

### LINKED ISSUE #127: Train default BertMesh
https://www.biorxiv.org/content/10.1101/2020.07.04.187674v1.full.pdf

### PR DESCRIPTION
**Pull request by nsorros**
## Description

Fixes #127 

## Checklist

- [x] Linked to Notion or GitHub issue
- [ ] Added tests
- [ ] Updated README
- [ ] DVC up to date

### CODE REVIEW

File: grants_tagger/bertmesh/prepare_data.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        label_binarizer = MultiLabelBinarizer(sparse_output=True)
+        label_binarizer.fit(Y)
+    
+    print("transforming data")
+    Y_vec = label_binarizer.transform(Y)
+    print(f"   number of labels {Y_vec.shape[1]}")
+
+    tokenizer = BertTokenizerFast.from_pretrained("bert-base-uncased")
+    X_vec = tokenizer(X, truncation=True, padding=True)["input_ids"]
+    X_vec = np.array(X_vec)
```
Comment (pdan93) [inline grants_tagger/bertmesh/prepare_data.py:47]: why transform it to a np.array here?
you can do same `.shape` calls and save it also as a tensor.

File: grants_tagger/bertmesh/train_torch.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        loss = loss_fct(logits.view(-1, self.model.config.num_labels),
+                        labels.float().view(-1, self.model.config.num_labels))
+        return (loss, outputs) if return_outputs else loss
+
+def train_bertmesh(x_path, y_path, model_path, multilabel_attention:bool=False,
+        batch_size: int=256, learning_rate: float=1e-4,
+        epochs: int=5, pretrained_model="bert-base-uncased"):
+    dataset = MeshDataset(x_path, y_path)
+
+    model = AutoModelForSequenceClassification.from_pretrained(pretrained_model, num_labels=dataset.num_labels)
```
Comment (pdan93) [inline grants_tagger/bertmesh/train_torch.py:50]: did you check if this trains ok on GPU? 
(you might have to do `.to(device)` for the model and the inputs)

**Comment by nsorros** [comment]
As results are not submitted so far. The first Pubmed Bert ran gave 0.40 and with multi label attention 0.49.

**Comment by nsorros** [comment]
313E02B9-230D-4BA8-ADE4-1483938DAA99

File: grants_tagger/bertmesh/model.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        )
+        return torch.matmul(torch.transpose(attention_weights, 2, 1), x)
+
+
+class BertMesh(torch.nn.Module):
+    def __init__(self, pretrained_model, num_labels, multilabel_attention=False):
+        super().__init__()
+        self.pretrained_model = pretrained_model
+        self.num_labels = num_labels
+        self.multilabel_attention = multilabel_attention
```
Comment (pdan93) [inline grants_tagger/bertmesh/model.py:23]: is this line needed if you overwrite it in 2 lines below?

File: grants_tagger/bertmesh/model.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+        self.bert = AutoModel.from_pretrained(pretrained_model)  # 768
+        self.multilabel_attention = MultiLabelAttention(
+            768, num_labels
+        )  # num_labels, 768
+        self.linear_1 = torch.nn.Linear(768, 512)  # num_labels, 512
```
Comment (pdan93) [inline grants_tagger/bertmesh/model.py:29]: consider using the sequential api for this?
```nn.Sequential(...)```

File: grants_tagger/bertmesh/train.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+from transformers import TFBertModel, AutoModel
+import tensorflow_addons as tfa
+import tensorflow as tf
+import scipy.sparse as sp
+import numpy as np
+import typer
+
+
+class MultiLabelAttention(tf.keras.layers.Layer):
```
Comment (pdan93) [inline grants_tagger/bertmesh/train.py:13]: should this be moved in a file on it's own ?

File: grants_tagger/bertmesh/prepare_data.py (hunk 1 above)
Comment (nsorros) [reply grants_tagger/bertmesh/prepare_data.py:61]: i guess so that we can use it both for tensorflow and pytorch training

File: grants_tagger/bertmesh/train.py (hunk 5 above)
Comment (nsorros) [reply grants_tagger/bertmesh/train.py:13]: possibly but i might delete the tensorflow train bit anyway. should i keep it? i mean most models are avaiable in pytorch not tensorflow. maybe i leave it and rename it to `train_tf` so the main train is the torch one

File: grants_tagger/bertmesh/model.py (hunk 4 above)
Comment (nsorros) [reply grants_tagger/bertmesh/model.py:29]: 👍

File: grants_tagger/bertmesh/model.py (hunk 3 above)
Comment (nsorros) [reply grants_tagger/bertmesh/model.py:23]: 🐛

File: grants_tagger/bertmesh/model.py (hunk 4 above)
Comment (nsorros) [reply grants_tagger/bertmesh/model.py:29]: actually in this case we use different layers depending on whether we add a multilabel attention or not. not all are used so this reduces duplication but generally good shout

File: grants_tagger/bertmesh/model.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        if self.multilabel_attention:
+            hidden_states = self.bert(input_ids=inputs)[0]
+            attention_outs = self.multilabel_attention(hidden_states)
+            outs = torch.nn.functional.relu(self.linear_1(attention_outs))
+            outs = torch.sigmoid(self.linear_2(outs))
+            outs = torch.flatten(outs, start_dim=1)
+        else:
+            cls = self.bert(input_ids=inputs)[1]
+            outs = torch.nn.functional.relu(self.linear_1(cls))
+            outs = torch.nn.functional.relu(self.linear_2(outs))
```
Comment (nsorros) [inline grants_tagger/bertmesh/model.py:43]: this should be remove as it is only relevant for multilabel attention

**Comment by nsorros** [comment]
Opening this PR for review. Note that flake8 error has surfaced a bug 🐛 to how I evaluate snapshot ensemble. 

The best result is achieved in commit 4c2f68ee9b674b0a53d60d6d396878c1b327f979

**Comment by aCampello** [comment]
Results look good, would be good to include time to train.

**Comment by aCampello** [comment]
What's your intention with this PR - merge as is? Or will you work more on that still?

**Comment by nsorros** [comment]
> What's your intention with this PR - merge as is? Or will you work more on that still?

This should be merged as is, it represents all the experiments I have run so far. It is experimental and not production ready i.e. there is no class abstraction that represents BertMesh and the code duplicates some things but it serves the purpose of moving fast towards trying a couple of ideas from the literature. It should also not be that far from being production ready anyway.

File: grants_tagger/bertmesh/results.json [hunk 7]
```diff
@@ -0,0 +1 @@
+{"p": 0.6955707011887685, "r": 0.569327457868485, "f1": 0.626149221959104, "th": 0.5}
```
Comment (aCampello) [inline grants_tagger/bertmesh/results.json:1]: It'd be nice to have train information such as time to train if I want to reproduce.

**Review by aCampello — APPROVED** [review]
I had read this PR before, don't know why I didn't approve.

How long did it take to train? I am keen to try and reproduce this with me AWS account.

File: grants_tagger/bertmesh/results.json (hunk 7 above)
Comment (nsorros) [reply grants_tagger/bertmesh/results.json:1]: Sure it can be added, I was not using our main train script to accelerate experimentation but I can add it easily.

**Comment by nsorros** [comment]
> I had read this PR before, don't know why I didn't approve.
> 
> How long did it take to train? I am keen to try and reproduce this with me AWS account.

Well actually there is a good opportunity to do this as soon as I change the model to produce a hugging face model which will make uploading to a hub super easy. So let me open this PR today and you can run it. It takes ~4 days in `g4dn.metal`. You can track progress with wandb
